### ISSUE TITLE
Try to position editor mouseover hint boxes so they never overlay current caret position region

### ISSUE THREAD

**Issue by rotemdan** [microsoft/vscode#19832]
This is related to #14840 but more of an enhancement suggestion on how to solve the issue without disabling hints completely.

This is something that can become a real annoyance at times, especially if the hint boxes become large, (in TypeScript, this happens frequently due to complex types).

Let's say I'm currently editing a source file:

![clipboard01](https://cloud.githubusercontent.com/assets/8589488/22590272/0b22839c-ea17-11e6-9b99-b5fef37fcb80.png)

Now I unintentionally moved the mouse cursor to hover over `options`, which has a complex interface type. When the cursor is positioned above it, a hint is shown:

![clipboard02](https://cloud.githubusercontent.com/assets/8589488/22590358/7ae6341c-ea17-11e6-854c-60d6f5608dd7.png)

Unfortunately the hint is large and is positioned over the area I'm currently editing. Usually I'm concentrated on editing and don't really realize why that happened so I just press the `Esc` key to get rid of it. However, even if I do that, but then I move the cursor again just a bit (sometimes unintentionally), the hint returns, and that becomes annoying at times.

My suggestion is to always position the hint box so it doesn't overlay some rectangle within the cursor's area. The logic is quite simple, in essence:

* If the caret is above the cursor, open the hint to the bottom.
* If the caret is below, open the hint to the top.
* If the caret is to the left, open the hint to the right.
* If the caret is to the right, open the hint to the left.

Of course this only matters if the caret is anywhere close to the cursor and the hint box is large enough to reach it, and there can be many potential adjustments and tweaks for various cases. I'm just giving this as an example.

**Comment by rotemdan** [comment]
Seems like there's another trigger for this: Auto Save.

If the cursor hovers an entity that triggers a hint, and text is typed. The behavior seems to be different when Auto Save is enabled or disabled (even without the mouse cursor being moved):

* When Auto Save is disabled and text is typed, the pop-up seems to disappear permanently.
* When Auto Save is enabled the popup seems to disappear for a second or two and then return, this happens every keystroke.

I have tested this several times on the same elements of the code repeatedly.

**Comment by alexandrudima** [comment]
fyi @bpasero regarding the Auto Save

**Comment by bpasero** [comment]
I cannot reproduce

**Comment by rotemdan** [comment]
@bpasero 

I have attempted to reproduce the particular auto-save behavior in `1.12.2` (Windows 10 x64) and did not succeed so far, maybe something has changed since the version I originally reported for.

Anyway, the general behavior I described looks like this:

![example1](https://cloud.githubusercontent.com/assets/8589488/26719698/331a65de-478e-11e7-8680-8a865c5ea225.gif)

The mouse cursor is hovering a target that triggers a tooltip. Even when using the arrow keys to move the caret (or typing) in text positioned underneath the tooltip, the tooltip disappears and quickly appears again after every key press. 

I remember noticing this didn't happen when auto-save was disabled, but that doesn't seem to be the case anymore or I was testing it differently, I'm not sure.

**Comment by ivolucien** [comment]
The issue in the original post still exists, I suggest that the blinking dialog issue should be a different ticket.

Mouse hover tips always appear immediately above the text in question, if you are editing text where the popup appears the text that you are editing is occluded by the tool tip window.

**Comment by alexandrudima** [comment]
We now have `"editor.hover.sticky"` which can be set to `false` which will make the hover go away when moving the mouse over it.

**Comment by ivolucien** [comment]
This ticket has been closed with a suggestion that does not address the original / primary issue for this ticket.

The `editor.hover.sticky` workaround does not address the request that hover tips avoid the area around the text cursor, and workaround for the side issue (persistent tip blocking edited text) is not very helpful.

Please re-open this ticket. 🙏

The goal of the request is IMO being able to edit text without having to take action to dismiss the hover tip.

Do you agree @rotemdan?

**Comment by rotemdan** [comment]
I submitted the issue in part because the tooltip appeared over the caret position and persisted (with some flicker) when text was written. This behavior seems to have been changed since (the tooltip no longer reappears after a key is pressed). This change (which doesn't seem to be related to `editor.hover.sticky`) has partially addressed the problem, I guess, at least in practice.

I've now tried `editor.hover.sticky` but it (subjectively) didn't seem to make much difference to me. I still feel having the tooltip positioned to minimize potential occlusion would have been a nice minor feature, at least as an option.
